Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1423 @@
+          .first { isIdentityOrCompatibleEncoding($0) }
+          .flatMap { $0 == .identity ? nil : $0 }
+      }
+
+      self.state = .clientOpenServerIdle(
+        .init(
+          previousState: state,
+          compressionAlgorithm: outboundEncoding,
+          decompressionConfiguration: .decompression(inboundEncoding)
+        )
+      )
+
+      return .receivedMetadata(Metadata(headers: metadata))
+    case .clientOpenServerIdle, .clientOpenServerOpen, .clientOpenServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Client shouldn't have sent metadata twice."
+      )
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Client can't have sent metadata if closed."
+      )
+    }
+  }
+
+  private mutating func serverReceive(bytes: ByteBuffer, endStream: Bool) throws {
+    switch self.state {
+    case .clientIdleServerIdle:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Can't have received a message if client is idle."
+      )
+    case .clientOpenServerIdle(var state):
+      // Deframer must be present on the server side, as we know the decompression
+      // algorithm from the moment the client opens.
+      assert(state.deframer != nil)
+      try state.deframer!.process(buffer: bytes) { deframedMessage in
+        state.inboundMessageBuffer.append(deframedMessage)
+      }
+
+      if endStream {
+        self.state = .clientClosedServerIdle(.init(previousState: state))
+      } else {
+        self.state = .clientOpenServerIdle(state)
+      }
+    case .clientOpenServerOpen(var state):
+      try state.deframer.process(buffer: bytes) { deframedMessage in
+        state.inboundMessageBuffer.append(deframedMessage)
+      }
+
+      if endStream {
+        self.state = .clientClosedServerOpen(.init(previousState: state))
+      } else {
+        self.state = .clientOpenServerOpen(state)
+      }
+    case .clientOpenServerClosed:
+      // Client is not done sending request, but server has already closed.
+      // Ignore the rest of the request: do nothing.
+      ()
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError(
+        "Client can't send a message if closed."
+      )
+    }
+  }
+
+  private mutating func serverNextOutboundMessage() throws -> OnNextOutboundMessage {
+    switch self.state {
+    case .clientIdleServerIdle, .clientOpenServerIdle, .clientClosedServerIdle:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError("Server is not open yet.")
+    case .clientOpenServerOpen(var state):
+      let response = try state.framer.next(compressor: state.compressor)
+      self.state = .clientOpenServerOpen(state)
+      return response != nil ? .sendMessage(response!) : .awaitMoreMessages
```

Comment:
nit: avoid the `!` by using `map`: `response.map { .sendMessage($0) } ?? .awaitMoreMessages`